### PR TITLE
SE: Learn NumberConstraint ranges for 2nd visits of conditions

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Follow-up to #7797
By applying the special treatment for binary number operations to not just loop conditions but also operations inside a loop, we simplify the code and fix some arcane FPs.

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs [hunk 1]
```diff
@@ -335,6 +335,22 @@ public void InfiniteForLoop(string arg)
             }
         }
 
+        public void LoopWithIfsInside(bool condition)
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:338]: Are those FPs due to we visited the loop only 2 times?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs [hunk 2]
```diff
@@ -335,6 +335,22 @@ public void InfiniteForLoop(string arg)
...
+        {
+            var done = false;
+            var i = 0;
+            while (!done)
+            {
+                if (i <= 5)                 // Noncompliant FP
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:344]: Just to verify my understanding:
this will reach learning that i: [1,5] after the max loop so it will learn that it can never enter the else.
Thus this is an FP due to not visiting the loop more.
Is this correct?

**Review by mary-georgiou-sonarsource — COMMENTED** [review]
To LGTM.
@pavel-mikula-sonarsource  please take a look.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs (hunk 2 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:344]: Yes. This is a bit unfortunate. But I don't see a good way to avoid these kinds of FPs without iterating loops more than twice.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs [hunk 3]
```diff
@@ -343,6 +343,22 @@ public void InfiniteForLoop(string arg)
             }
         }
 
+        public void LoopWithIfsInside(bool condition)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:346]: I would like to see some positive impact of this change to a rule. Otherwise, there's not really a reason to do it.

What problem does it actually solve?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:352]: Was this already problem before (likely yes), or was it created with this PR?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 4]
```diff
@@ -219,6 +215,16 @@ private static SymbolicConstraint BinaryConstraint(BinaryOperatorKind kind, Symb
         }
     }
 
+    private static BoolConstraint BinaryNullableBoolConstraint(BinaryOperatorKind kind, BoolConstraint boolConstraint, bool oneIsNull) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:218]: `oneIsNull` makes sense in scope of this refactoring, but will be hard to understand without the refactoring context.
I also wouldn't mind passing both values in, as `when leftIsNull || rightIsNull` is very clear to understand
```suggestion
    private static BoolConstraint BinaryNullableBoolConstraint(BinaryOperatorKind kind, BoolConstraint boolConstraint, bool hasNull) =>
```

Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:218]: The `bool` doesn't help here. More the opposite.
```suggestion
    private static BoolConstraint BinaryNullableBoolConstraint(BinaryOperatorKind kind, BoolConstraint constraint, bool oneIsNull) =>
```

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM with some questions to look into

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs (hunk 2 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:352]: This was an issue before this PR already.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs (hunk 3 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:346]: The impact is minimal. If you review by commit, you can see that one edge-case FP was fixed. The main reason is to simplify the code in `Binary.cs`.

**Comment by pavel-mikula-sonarsource** [comment]
It would have been better to merge them master with rebase. To preserve the history, smaller changes and PR comments for archeology. It's going to be bit harder to discover this one now
